server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -44,12 +45,14 @@ final class FieldTypeLookup {
     ) {
 
         final Map<String, MappedFieldType> fullNameToFieldType = new HashMap<>();
+        final Map<String, String> fullSubfieldNameToParentPath = new HashMap<>();
```

Comment:
Rather than adding an extra map here I think we can extend the use of `fieldToCopiedFields`.  Multifields are essentially a specialised version of copy_to fields after all.

Reply:
And maybe rename it to something like `sourceLocations` so that it covers both copy_to and multi_field.

Reply:
@romseygeek I am not sure how nit this would be...
If field `A` is copied to `B`, then we want to be able to tell that `A` is a valid source path for `B`, as well as `B`. If `B` also has a multifield mapping to `B.subfield`, we want to be able to tell that the source paths to `B.subfield` are all `B` source paths. While it's trivial to tell that `B` is a source path for `B.subfield` when going through all mappers, it is not as trivial to tell that `A` is a source path as well. 

To demonstrate, assume that these mappings are iterated in the following order:
1. field `B` mapper is analyzed: 
   - we add a mapping: `B.subfield --> [B]`
2. field `A` mapper is analyzed: 
   - we add a mapping: `B --> [B, A]`
   - we also need to update the former mapping: `B.subfield --> [B, A]`
For every `copy_of`, we would now need to go over all values of the new combined map and look for the `copy_of` source field and update the set with the target field if we find such.

While this is definitely doable, it is much less straightforward than the current algorithm, which does't try to maintain copies of same sets for multifields and instead only [resolve this in lookup time](https://github.com/elastic/elasticsearch/blob/d36f889003fa39339c9f296117fe5223b4dbe1b8/server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java#L200-L205).

Therefore, I still recommend we stick with the additional mapping, but if this doesn't convince you, I will gladly change.